### ISSUE TITLE
Unit testing

### ISSUE THREAD

**Issue by bingao** [openrsp/openrsp#6]
If time allowed, make the unit testing easier and optionally add more unit tests before release.

**Comment by bast** [comment]
For this I will need some information on how to build and run the unit tests right now. Then I can hook it up with CMake. I know the documentation is somewhere but collecting the pointers here in the issue with help.

**Comment by bast** [comment]
Unassigning myself: I like to self-assign once I really start working on it.

**Comment by bingao** [comment]
Hei @bast 

The unit testing needs to compile QcMatrix first. Do you think if it is a good idea to include QcMatrix as an external repository into OpenRSP? Or do you have any other better solution?

Thank you.

**Comment by bast** [comment]
Yes, we can add it as external dependency but it does not have to be a submodule. IMO it should not. Can you describe here how you would fetch and build the dependency and run the tests on the command line? Then I can translate it to a Travis CI recipe.

**Comment by bingao** [comment]
Hei @bast 

I am not sure if it is able to build using Travis CI. Anyway, here are the steps:

1. One first needs to clone the QcMatrix library,
`git clone https://gitlab.com/bingao/qcmatrix.git`

2. Switch to a workable commit,
`cd qcmatrix`
`git checkout 77529cd2705e76ea5f5635a78b45f2eaf361ea4d`

3. Build the QcMatrix library, for instance,
`mkdir build`
`cd build`
`cmake .. -DQCMATRIX_ENABLE_VIEW=F03 -DQCMATRIX_ENABLE_VIEW=ON -DQCMATRIX_TEST_EXECUTABLE=OFF -DQCMATRIX_Fortran_API=F03`

4. After successfully building QcMatrix, one needs to set up OpenRSP using CMake as,
`cmake path-to-openrsp-cmakelist.txt -DOPENRSP_TEST_EXECUTABLE=ON -DOPENRSP_FORTRAN_API=ON -DQCMATRIX_HEADER_DIR="path-to-qcmatrix/include;path-to-qcmatrix/build" -DQCMATRIX_MODULE_DIR=path-to-qcmatrix/build -DQCMATRIX_LIB=path-to-qcmatrix/build/libqcmatrix.a`

The problem is we need to find where the BLAS and LAPACK libraries are, and append such information to, for example, `-DQCMATRIX_LIB="path-to-qcmatrix/build/libqcmatrix.a;blas-and-lapack-libraries"`.

Do you think if we can use Travis CI to do the above 4 steps? Thank you.

**Comment by bast** [comment]
I am very confident that we can get it done with Travis. Thanks for the explanations. I will have a go at it and ask if I get stuck.

**Comment by bingao** [comment]
Great. Thank you.

**Comment by bast** [comment]
The qcmatrix library in the above description is not building the `qcmatrix_f.mod`, probably some option is missing?

**Comment by bast** [comment]
Also if you try to recreate the above, there is a typo: `-DDQCMATRIX_TEST_EXECUTABLE` (repeated "D").

**Comment by bast** [comment]
Perhaps OpenRSP should not be looking for `qcmatrix_f.mod` but then some other options is perhaps missing.

**Comment by bingao** [comment]
Hei @bast 

Sorry that I forgot one also need `-DQCMATRIX_Fortran_API=F03` in the step 3.

I add it into my previous comment and remove the duplicated `D`. Thank you.

**Comment by bast** [comment]
Thanks!

**Comment by bingao** [comment]
> Perhaps OpenRSP should not be looking for `qcmatrix_f.mod` but then some other options is perhaps missing.

Yes. I also feel it is weird, because the variable `OPENRSP_FORTRAN_API` in OpenRSP `CMakeLists.txt` is `OFF`, and it should not look for `qcmatrix_f.mod`. I will look into later.

**Comment by bast** [comment]
I managed to get it almost working but now getting:
```
undefined reference to `lsquit_'
```
Is this expected? How do you work around that?

**Comment by bast** [comment]
Full output: https://travis-ci.org/openrsp/openrsp/builds/442622038

**Comment by bingao** [comment]
Hei @bast 

Yes, I fixed this bug. I am pushing it now, and write to you once it is done.

**Comment by bast** [comment]
The build and test is implemented in #15.

**Comment by bast** [comment]
Why is it reopened? Fixing the test is a separate issue.
